### PR TITLE
Implements .samrc configuration file

### PR DESCRIPTION
**Pull request by heitorlessa**
*Issue #, if available:*

*Description of changes:*

This provides SAM CLI with a `.samrc` configuration file allowing users to create either project-based or user's home folder configuration.

**Implemented**

* [x] Initial Config implementation (load, parsing)
* [x] Merge user and project configuration
* [x] samrc singleton
* [x] Functional tests
* [x] Config validation
* [x] Basic schema for required and optional field values
* [x] Debug messages
* [x] Unit tests
* [x] Address Jacob's feedback and simplify implementation

**Pending**

* [x] Agree on Singleton vs Getter for exposing config within SAM CLI
* [x] Implement Config in one of the commands available
* [x] Document what .samrc should look and how it should be created from customers
* [ ] Patch from Jacob to use click.get_app_dir and other fixes for app dir

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samcli/config/config.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        """
+
+        LOG.debug("Validating SAMRC config with given JSONSchema")
+        jsonschema.validate(config, schema)
+
+        LOG.debug("SAMRC looks valid!")
+        LOG.debug("Schema used: %s", schema)
+        LOG.debug("Config used: %s", config)
+
+        return True
```
Comment (jfuss) [inline samcli/config/config.py]: When is this false? Does `jsonschema.validate` through validation exceptions? 

The return here is used, consider removing.

File: samcli/config/config.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    def merge_config(self, user_config, project_config):
+        """Merge project and user configuration into a single dictionary
+
+        Creates a new configuration with both configuration merged
+        it favours project level over user configuration if keys are duplicated
+
+        NOTE
+        ----
+            It takes any number of nested dicts
+            but it doesn't take lists into account
```
Comment (jfuss) [inline samcli/config/config.py]: What does we do for lists?

File: samcli/config/config.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+        # Project may have unique config we need to copy over too
+        # so that we can have user+project config available as one
+        for key in project_config:
+            if key not in user_config:
+                user_config[key] = project_config[key]
+
+        return user_config
+
+    def __has_user_config(self):
```
Comment (jfuss) [inline samcli/config/config.py]: `__has_user_config` and `__has_project_config` are the same methods but do the operation on different files. These can be one method that accept some path to check.

File: samcli/config/config.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
+"""
+Config module that reads .samrc definition
+"""
+
+import logging
+import json
+
+import jsonschema
```
[the remark this answers is not in the document]
Comment (heitorlessa) [reply samcli/config/config.py]: Re-added jsonschema under base.txt

**Comment by jfuss** [comment]
@heitorlessa Are you still working through these changes? Do you need any help?

File: samcli/config/config.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        -------
+        Dict
+            Parsed YAML configuration as dictionary
+        """
+        config = Path(config)
+
+        return yaml.safe_load(config.read_text())
+
+
+samrc = Config().load()
```
[the remark this answers is not in the document]
Comment (heitorlessa) [reply samcli/config/config.py]: These variables[1] were requested by @sanathkr for future use in other commands (from samcli.config.config import samrc; samrc['default']...).

That came from this singleton: https://github.com/python/cpython/blob/master/Lib/logging/__init__.py#L1892

If my memory serves we chatted on adding a getter function within `config/__init__.py` to return config instead:

**Getter**

```python
# Getter in samcli/config/__init__.py
from samcli.config.config import Config

def get_config():
    return Config().load()
```

**Usage**

```python
# Somewhere in SAM CLI
import samcli.config as config  

config = config.get_config()
....
```

Let me know the consensus there and I'm happy to address that

File: samcli/config/config.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        Tuple
+            Tuple with both configs and whether they were found
+
+        Example
+        -------
+            > user_config, project_config = self.__find_config()
+        """
+
+        self.__user_config_file = Path.home().joinpath('.samrc')
+        self.__project_config_file = Path.cwd().joinpath('.samrc')
```
[the remark this answers is not in the document]
Comment (heitorlessa) [reply samcli/config/config.py]: We agreed to have a patch in here as you already implemented a better version somewhere; leaving this unresolved for now

File: samcli/config/config.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+        if self.__has_user_config() and self.__has_project_config():
```
[the remark this answers is not in the document]
Comment (heitorlessa) [reply samcli/config/config.py]: Now simplified

File: samcli/config/config.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        Returns
+        -------
+        Dict
+            SAMRC configuration
+        """
+
+        LOG.debug("Looking for SAMRC before attempting to load")
+        user_config, project_config = self.__find_config()
+
+        if user_config is not None and project_config is not None:
```
[the remark this answers is not in the document]
Comment (heitorlessa) [reply samcli/config/config.py]: Now simplified

File: samcli/config/config.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+            LOG.debug("%s", user_config)
+            self.config = self.__read_config(user_config)
+        elif project_config is not None and user_config is None:
+            LOG.debug("Found one SAMRC -- Project specific")
+            LOG.debug("%s", project_config)
+            self.config = self.__read_config(project_config)
+
+        LOG.debug("SAMRC is ready to be validated")
+        LOG.debug("%s", self.config)
+        self.validate_config(self.config, self.schema)
```
[the remark this answers is not in the document]
Comment (heitorlessa) [reply samcli/config/config.py]: Addressed

File: samcli/config/config.py (hunk 2 above)
Comment (heitorlessa) [reply samcli/config/config.py]: Addressed

File: samcli/config/config.py (hunk 3 above)
Comment (heitorlessa) [reply samcli/config/config.py]: Addressed

**Comment by heitorlessa** [comment]
@jfuss Apologies for the delay. Pushed the latest local changes I had that addressed all major comments - Here's what pending:

* [ ] Agree on Singleton vs Getter for exposing config within SAM CLI
* [ ] Implement Config in one of the commands available
* [ ] Document what .samrc should look and how it should be created from customers
* [ ] Patch from Jacob to use click.get_app_dir and other fixes for app dir

**Comment by heitorlessa** [comment]
Pushed the last one and only pending Patch at this stage @jfuss.

**Comment by jfuss** [comment]
Capturing some thoughts:

1. We need to update the Manifest.in file to include the new .json file. Otherwise when we go to create a release, the schema.json that the SAMRC needs will be missing.
2. We need to vend in the current working directory. Right now we get the cwd by Path().cwd() but the rest of the commands use the cwd being the directory the template.yml file is.

**Comment by heitorlessa** [comment]
@jfuss added `schema.json` under `Manifest.in` and generated a local distribution to confirm:

```bash
python setup.py sdist
find dist/ -type f -name 'schema.json'
./samcli/config/schema.json
```

**Comment by jfuss** [comment]
`.samrc` needs a little more thought into it before going further here. I want to abstract away a lot of the paths and some other content so we have a clean implementation that works for CLI. I started a design doc for this, which I am hoping to pick up here soon.

**Comment by heitorlessa** [comment]
👍

**Comment by sanathkr** [comment]
Closing this PR for now because this is pending design #975. Most of the implementation here is great. So when we have a solid design we can revive this PR.

**Pull request closed without merging**